src/pyclass.rs

```diff
@@ -0,0 +1,642 @@
+use std::ptr::{self, NonNull};
+
+#[inline]
+pub(crate) unsafe fn default_alloc<T: PyTypeInfo>() -> *mut ffi::PyObject {
+    let tp_ptr = T::type_object();
+    if T::FLAGS & type_flags::EXTENDED != 0
+        && <T::BaseType as PyTypeInfo>::ConcreteLayout::IS_NATIVE_TYPE
+    {
+        let base_tp_ptr = <T::BaseType as PyTypeInfo>::type_object();
+        if let Some(base_new) = (*base_tp_ptr).tp_new {
+            return base_new(tp_ptr, ptr::null_mut(), ptr::null_mut());
+        }
+    }
+    let alloc = (*tp_ptr).tp_alloc.unwrap_or(ffi::PyType_GenericAlloc);
+    alloc(tp_ptr, 0)
+}
+
+/// A trait that enables custome alloc/dealloc implementations for pyclasses.
+pub trait PyClassAlloc: PyTypeInfo + Sized {
+    unsafe fn alloc(_py: Python) -> *mut Self::ConcreteLayout {
+        default_alloc::<Self>() as _
+    }
+
+    unsafe fn dealloc(py: Python, self_: *mut Self::ConcreteLayout) {
+        (*self_).py_drop(py);
+        let obj = self_ as _;
+        if ffi::PyObject_CallFinalizerFromDealloc(obj) < 0 {
+            return;
+        }
+
+        match Self::type_object().tp_free {
+            Some(free) => free(obj as *mut c_void),
+            None => tp_free_fallback(obj),
+        }
+    }
+}
+
+#[doc(hidden)]
+pub unsafe fn tp_free_fallback(obj: *mut ffi::PyObject) {
+    let ty = ffi::Py_TYPE(obj);
+    if ffi::PyType_IS_GC(ty) != 0 {
+        ffi::PyObject_GC_Del(obj as *mut c_void);
+    } else {
+        ffi::PyObject_Free(obj as *mut c_void);
+    }
+
+    // For heap types, PyType_GenericAlloc calls INCREF on the type objects,
+    // so we need to call DECREF here:
+    if ffi::PyType_HasFeature(ty, ffi::Py_TPFLAGS_HEAPTYPE) != 0 {
+        ffi::Py_DECREF(ty as *mut ffi::PyObject);
+    }
+}
+
+/// If `PyClass` is implemented for `T`, then we can use `T` in the Python world,
+/// via `PyClassShell`.
+///
+/// `#[pyclass]` attribute automatically implement this trait for your Rust struct,
+/// so you don't have to use this trait directly.
+pub trait PyClass:
+    PyTypeInfo<ConcreteLayout = PyClassShell<Self>> + Sized + PyClassAlloc + PyMethodsProtocol
+{
+    type Dict: PyClassDict;
+    type WeakRef: PyClassWeakRef;
+}
+
+unsafe impl<T> PyTypeObject for T
+where
+    T: PyClass,
+{
+    fn init_type() -> NonNull<ffi::PyTypeObject> {
+        let type_object = unsafe { <Self as PyTypeInfo>::type_object() };
+
+        if (type_object.tp_flags & ffi::Py_TPFLAGS_READY) == 0 {
+            // automatically initialize the class on-demand
+            let gil = Python::acquire_gil();
+            let py = gil.python();
+
+            initialize_type::<Self>(py, <Self as PyTypeInfo>::MODULE).unwrap_or_else(|e| {
+                e.print(py);
+                panic!("An error occurred while initializing class {}", Self::NAME)
+            });
+        }
+
+        unsafe { NonNull::new_unchecked(type_object) }
+    }
+}
+
+/// `PyClassShell` represents the concrete layout of `PyClass` in the Python heap.
```

Comment:
Is `PyClass` another class, or are you saying that this represents the concrete layout of *a Python class* on the Python heap? If the latter, I would reword it like this:

``` `PyClassShell` represents the concrete layout of a Python class on the Python heap```

(Maybe you can exclude the "Python" before `class`).